abracadabra/src/writer/hdf5_writer.cc

```diff
@@ -0,0 +1,82 @@
+#include "hdf5_writer.hh"
+
+#include <cstring>
+
+
+hdf5_writer::hdf5_writer():
+    file_(0), irun_(0), iparticle_(0) {}
+
+
+void hdf5_writer::open(std::string file_name) {
+    file_ = H5Fcreate(file_name.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+
+    std::string group_name = "/MC";
+    group_ = create_group(file_, group_name);
+
+    std::string run_table_name = "configuration";
+    memtype_run_ = create_run_type();
+    run_table_   = create_table(group_, run_table_name, memtype_run_);
+
+    std::string particle_table_name = "particles";
+    memtype_particle_ = create_particle_type();
+    particle_table_ = create_table(group_, particle_table_name, memtype_particle_);
+
+}
+
+void hdf5_writer::write_run_info(const char* param_key, const char* param_value) {
+    run_info_t run_data;
+    memset(run_data.param_key,   0, CONFLEN);
+    memset(run_data.param_value, 0, CONFLEN);
+    strcpy(run_data.param_key, param_key);
+    strcpy(run_data.param_value, param_value);
+    write_table_data((void*) &run_data, run_table_, memtype_run_, irun_);
+
+    irun_++;
+}
+
+
+void hdf5_writer::write_particle_info(int evt_id, trajectory * trj, bool primary, int mother_id, float kin_energy) {
+    particle_t particle;
+
+    particle.event_id    = evt_id;
+    particle.particle_id = trj->GetTrackID();
+    particle.primary     = (char) primary;
+    particle.mother_id   = mother_id;
+    particle.kin_energy  = kin_energy;
+    particle.length      = trj->GetTrackLength();
+
+    // Initialize strings with 0's
+    memset(particle.particle_name , 0, STRLEN);
+    memset(particle.initial_volume, 0, STRLEN);
+    memset(particle.final_volume  , 0, STRLEN);
+    memset(particle.creator_proc  , 0, STRLEN);
+    memset(particle.final_proc    , 0, STRLEN);
+
+    // Copy actual values
+    strcpy(particle.particle_name , trj->GetParticleName()  .c_str());
+    strcpy(particle.initial_volume, trj->GetInitialVolume() .c_str());
+    strcpy(particle.final_volume  , trj->GetFinalVolume()   .c_str());
+    strcpy(particle.creator_proc  , trj->GetCreatorProcess().c_str());
+    strcpy(particle.final_proc    , trj->GetFinalProcess()  .c_str());
```

Comment:
Create (and use) an abstraction for this pattern. Some variation on the theme of:

``` c++
void set_string(char * const to, std::string const& from) {
  memset(to, 0, STRLEN);
  strcpy(to, from.c_str());
}
```

and, whilst we're at it, probably do

``` c++
void set_param(char * const to, char const * const from) {
  memset(to, 0, CONFLEN);
  strcpy(to, from);
}
```

further up, as well.

Reply:
Done.